pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -148,8 +152,32 @@ public synchronized void registerSendOp(CompletableFuture<MessageId> sendFuture)
         }
     }
 
+    public synchronized <T> void registerInternal(CompletableFuture<T> completableFuture) {
+        //There have been an exception, it means this transaction should not be commit,
+        //so there is no need to record the execution of other operations
+        if (!executedFuture.isCompletedExceptionally()) {
+            opsExecutingInTxn.incrementAndGet();
+            executedFuture = new CompletableFuture<>();
+            completableFuture.thenRun(() -> {
+                synchronized (this) {
```

Comment:
Why need `synchronized` here?

Reply:
`executedFuture = new CompletableFuture<>();`, `executedFuture.complete(null);` and `executedFuture.completeExceptionally` should be exclusive.